src/essreflectometry/orso.py

```diff
@@ -1,15 +1,173 @@
+"""ORSO measurement."""
+
+
+def parse_orso_experiment(raw_data: RawData[Run]) -> OrsoExperiment[Run]:
+    """Parse ORSO experiment metadata from raw NeXus data."""
+    return OrsoExperiment(
+        data_source.Experiment(
+            title=raw_data['title'],
+            instrument=raw_data['name'],
+            facility=raw_data['facility'],
+            start_date=parse_datetime(raw_data['start_time']),
+            probe='neutron',
+        )
+    )
+
+
+def parse_orso_owner(raw_data: RawData[Run]) -> OrsoOwner[Run]:
+    """Parse ORSO owner metadata from raw NeXus data."""
+    return OrsoOwner(
+        orso_base.Person(
+            name=raw_data['user']['name'],
+            contact=raw_data['user']['email'],
+            affiliation=None,
+        )
+    )
+
+
+def parse_orso_sample(raw_data: RawData[Run]) -> OrsoSample[Run]:
+    """Parse ORSO sample metadata from raw NeXus data."""
+    if not raw_data.get('sample'):
+        return OrsoSample(data_source.Sample.empty())
+    raise NotImplementedError('NeXus sample parsing is not implemented')
+
+
+def build_orso_measurement(
+    sample_filename: Filename[Sample],
+    reference_filename: Optional[Filename[Reference]],
+    instrument: Optional[OrsoInstrument[Sample]],
+) -> OrsoMeasurement:
+    """Assemble ORSO measurement metadata."""
+    # TODO populate timestamp
+    #      doesn't work with a local file because we need the timestamp of the original,
+    #      SciCat can provide that
+    if reference_filename:
+        additional_files = [
+            orso_base.File(
+                file=os.path.basename(reference_filename), comment='supermirror'
+            )
+        ]
+    else:
+        additional_files = []
+    return OrsoMeasurement(
+        data_source.Measurement(
+            instrument_settings=instrument,
+            data_files=[orso_base.File(file=os.path.basename(sample_filename))],
+            additional_files=additional_files,
+        )
+    )
+
+
+def build_orso_reduction(creator: Optional[OrsoCreator]) -> OrsoReduction:
+    """Construct ORSO reduction metadata.
+
+    This assumes that ess.reflectometry is the primary piece of software
+    used to reduce the data.
     """
-    A function to raise a orso specific error if necessary.
+    # Import here to break cycle __init__ -> io -> orso -> __init__
+    from . import __version__
+
+    return OrsoReduction(
+        reduction.Reduction(
+            software=reduction.Software(
+                name='ess.reflectometry',
+                version=str(__version__),
+                platform=platform.system(),
+            ),
+            timestamp=datetime.now(tz=timezone.utc),
+            creator=creator,
+            corrections=[],
+        )
+    )
+
+
+def build_orso_data_source(
+    owner: Optional[OrsoOwner[Sample]],
+    sample: Optional[OrsoSample[Sample]],
+    sample_experiment: Optional[OrsoExperiment[Sample]],
+    reference_experiment: Optional[OrsoExperiment[Reference]],
```

Comment:
See above, I don't think the reference measurement is or should be an "experiment", i.e., I'd remove this.

Reply:
I think you use the opposite definition of 'experiment' and 'measurement' as ORSO. An experiment is one run at an instrument, i.e., sample and reference are separate experiments. And a measurement is a combination of any number of experiments.

But yes, we could ignore the reference experiment. `build_orso_measurement` already only includes the sample instrument.

Reply:
Can you provide a reference? Google found this: https://www.reflectometry.org/projects/file_formats/dictionaries/, and it seems to be consistent with "my" (or rather NeXus') definition.

Reply:
Similar here: https://www.reflectometry.org/projects/file_formats/tasks/ws_2021-06_text/

Reply:
The actual specification is here: https://github.com/reflectivity/file_format/blob/master/specification.md

But you are right. 'Experiment' refers to a 'series of measurements'. However, the 'measurement' key also refers to a series.